### ISSUE TITLE
Cleaning up the Parser: lit(), secure(), and escape()

### ISSUE THREAD

**Issue by captdeaf** [pennmush/pennmush#1217]
FN_LITERAL is parser brain damage that needs to go away. `lit()` is its sole culprit.

Also:
> Remove `secure()` and `escape()` - and/or make them both return `#-1 LEARN HOW THE PARSER F****** WORKS`.
> \- Abraham Lincoln \- Michael Scott

Unfortunately:
```
mush@gungnir:~/game/data$ grep -i 'lit(' indb | wc -l
511
mush@gungnir:~/game/data$ grep -i 'escape(' indb | wc -l
142
mush@gungnir:~/game/data$ grep -i 'secure(' indb | wc -l
176
```

My suggestion: Make `secure()` and `escape()` both aliases for `decompose()`, and marked deprecated.

`lit()` is somewhat trickier - skimming its uses on VISIBLE objects, it's used quite a lot with ASCII art.

I vote we just make `lit()` a NOPARSE clone of strcat.

**Comment by mike347** [comment]
I understand the passion behind wanting to get rid of these, but it will break backwards compatibility on older games.. like the one I admin on.

**Comment by talvo** [comment]
As much as I would genuinely like to remove secure() and escape(), which are misused far more than they're decently used, it would be a massive backwards-incompatability break (replacing with decompose() would be just as bad).

Making lit() NOPARSE would also break stuff (and make the thing even more useless than it is now); my only intended change for that was to stop "(" having any special meaning in FN_LITERAL.

**Comment by captdeaf** [comment]
Renamed the Issue topic to just cover `lit()`. I'd like to change `lit()` from `FN_LITERAL` to `FN_NOPARSE`

For the 'good' use cases of `lit()`, this would have no effect. Where this would have an effect is any `lit()` uses with mismatched `{}[]()` nesting.

e.g: `lit({)` would break. (Use `\{`, anyway!). `map(lit(#lambda/[ansi(%0,%0)),r g b)` would break because of the mismatched `[`.

To figure out what kind of impact this would have:

I ran a perl script (attaching to this issue) that tested `M*U*S*H`'s `data/outdb`, printing (only) the statistics:
```
Total instances of lit(): 749
Total instances of lit() with {}[]s inside: 58
Total instances of lit() with mismatched {}[]s inside: 15
```

A skim of the `lit()` with mismatched `[](){}`s that are on `VISUAL` objects (use `scanlit.pl -v`) - At least 5 of them are from `@clone`s of an old version of my calculator code that is long past broken, anyway :man_facepalming: . The others (again, still limited to visual) are broken or poorly formed bits of code, such as the above `#lambda/ansi` example, or regexps that still work from pure dumb luck.

But I think a measly 15, out of 749 uses of lit, in a code-heavy codebase as `M*U*S*H`, is a good sign it's okay to make `lit()` `NOPARSE` and get rid of `FN_LITERAL` for once and for all :D.

**Comment by captdeaf** [comment]
Would also appreciate the output from anyone else running `scanlit.pl data/indb`, in case the number varies wildly. (the .txt is required by github since it doesn't support .pl)

**Comment by mike347** [comment]
From ATS: 
```
Total instances of lit(): 424
Total instances of lit() with {}[]s inside: 49
Total instances of lit() with mismatched {}[]s inside: 23
```

**Comment by mike347** [comment]
So what you're saying is, there will only be issues with mismatched stuff?
Can your sprint tell me what objects/attrs have this issue?

**Comment by talvo** [comment]
Sorry, Walker, but this one's a no from me; changing it this much provides very little, but will potentially break a lot of stuff.

**Comment by captdeaf** [comment]
"provides very little" - it'll enable me to effect a parser rewrite, including building expression trees. Maybe even bytecode compiling for fun! =).

**Comment by talvo** [comment]
Provides very little to users; your fun in doing it doesn't count, I'm afraid. ;)

**Comment by MrWigggles** [comment]
So catty.

**Comment by shawnw** [comment]
Yeah, changing `lit()` to not allow unbalanced braces would be bad. Only I'm allowed to break backwards compatibility like that.
